clang, prohibit first level const in function signatures (#1733)
Azure/azure-sdk: docs/clang/design.md

{% include requirement/SHOULD id="clang-design-naming-paramnames" %} use a meaningful name for parameters and local variable names.  Parameters and local variable names should be named as all lower-case words, separated by underscores (snake-casing).

> {% include requirements/MUSTNOT id="clang-design-const-parameters" %} use first level const for parameters or return types. Pointers to const are fine. First level const on function parameters does not
> provide any additional grantees to the caller, and can be confusing. For example:

Reviewer: ```suggestion
{% include requirements/MUSTNOT id="clang-design-const-parameters" %} use first level const for parameters or return types. Pointers to const are fine. First level const on function parameters do not
provide any additional guarantees to the caller, and can be confusing. For example:
```

{% include requirement/SHOULD id="clang-design-naming-paramnames" %} use a meaningful name for parameters and local variable names.  Parameters and local variable names should be named as all lower-case words, separated by underscores (snake-casing).

> {% include requirement/MUSTNOT id="clang-design-const-parameters" %} use first level const for parameters or return types for function signatures. 

Reviewer: If we are sticking to defining/declaring, this should be:
```suggestion
{% include requirement/MUSTNOT id="clang-design-const-parameters" %} use first level const for parameters or return types for function declarations in headers. 
```

Pointers to const are fine. First level const on function parameters do not provide any additional guarantees to the caller, and can be confusing. For example:

This is not OK:
```c
const int az_iot_client_set_widget_id(az_iot_client *client, const int id);
```

Instead write:
```c
int az_iot_client_set_widget_id(az_iot_client *client, int id);
```

This is also not OK
```c
const int az_iot_client_get_widget_id(az_iot_client const *const client, int *const id);
```

Instead write:
```c
int az_iot_client_get(az_iot_client const *client, int *id);
```

> Note: You may use first level const when _defining_ a function, if you want to ensure you don't modify the value. Similarly you may use first level const on

Reviewer: Is defining vs. declaring established terms for src vs. header, respectively?

I always thought as those two being synonymous, and what's in source is referred to as "implementation".

If so,
```suggestion
Note: You may _optionally_ use first level const in the _implementation_ of a function, if you want to ensure you don't modify the value. Similarly, you may use first level const on
```

Reviewer: And if you prefer keeping defining to mean what's in source:
```suggestion
Note: You may _optionally_ use first level const in the _definition_ of a function, if you want to ensure you don't modify the value. Similarly, you may use first level const on
```

> definitions of inline functions (which are in headers).

Reviewer: ```suggestion
definitions of inline functions (which are declared in headers).
```
